main.py

```diff
@@ -0,0 +1,123 @@
+import os
+import urllib.parse
+import time
+import logging
+import re
+
+from github import Github
+from gitlab import Gitlab
+from prometheus_client import start_http_server
+from prometheus_client.core import REGISTRY
+from sentry_sdk import init as sentry_sdk_init
+
+from lib.saas_repo import get_saas_repos, SaasRepo
+from lib.repo import GHRepo, GLRepo, CommitNotFound
+from lib.metrics import SaasCollector
+from lib.gql import GqlApi
+
+logger = logging.getLogger()
+
+logging.basicConfig(format='%(levelname)s: %(message)s', level=logging.INFO)
+
+GITHUB_TOKEN = os.environ['GITHUB_TOKEN']
+GITLAB_SERVER = os.environ['GITLAB_SERVER']
+GITLAB_TOKEN = os.environ['GITLAB_TOKEN']
+
+APP_INTERFACE_URL = os.environ['APP_INTERFACE_URL']
+APP_INTERFACE_TOKEN = os.environ['APP_INTERFACE_TOKEN']
+
+SENTRY_DSN = os.getenv('SENTRY_DSN')
+
+gh_client = Github(GITHUB_TOKEN)
+gl_client = Gitlab(GITLAB_SERVER, private_token=GITLAB_TOKEN)
+gql_client = GqlApi(APP_INTERFACE_URL, APP_INTERFACE_TOKEN)
+
+if os.getenv('SLEEP_TIME'):
+    SLEEP_TIME = int(os.environ['SLEEP_TIME'])
+else:
+    SLEEP_TIME = 300
+
+
+if SENTRY_DSN:
+    def before_breadcrumb(crumb, hint):
+        # remove authentication from http urls
+        if crumb['category'] == 'subprocess':
+            crumb['message'] = re.sub(r'//.*?@', '//***@', crumb['message'])
+        return crumb
+
+    sentry_sdk_init(SENTRY_DSN, before_breadcrumb=before_breadcrumb)
+
+
+def inject_auth(url, auth):
+    parts = list(urllib.parse.urlsplit(url))
+    parts[1] = auth + '@' + parts[1]
+    return urllib.parse.urlunsplit(parts)
+
+
+def get_stats(saas_repos):
+    stats = []
+    for saas_repo_full in saas_repos:
+        logging.info(['saas_repo', saas_repo_full])
+
+        if saas_repo_full.startswith(GITLAB_SERVER):
+            saas_repo_auth = inject_auth(saas_repo_full, GITLAB_TOKEN)
+        else:
+            saas_repo_auth = saas_repo_full
```

Comment:
The token injection login should probably move into the `SaasRepo()` `__init__()` method.